File: Language/C#/Book/LINQ实战/第13章 应用程序各层中的LINQ/Chapter13.CSharp - LinqBooks/LinqBooks.ImportFromAmazon/ImportForm.cs
#region Imports

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Data.Linq;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.Xml.Linq;

using LinqBooks.Entities;
using AmazonCommon;

#endregion Imports
  
namespace LinqBooks.ImportFromAmazon
{
  public partial class ImportForm : Form
  {
    LinqBooksDataContext ctx = new LinqBooksDataContext(Properties.Settings.Default.liaConnectionString);
    XNamespace ns = Amazon.AmazonNS;
    XElement amazonXml;

    public ImportForm()
    {
      InitializeComponent();
      this.Load += new EventHandler(ImportForm_Load);
    }

    void ImportForm_Load(object sender, EventArgs e)
    {
      categoryComboBox.DataSource = ctx.Subjects.OrderBy(subject => subject.Name).ToList();
    }

    void searchButton_Click(object sender, EventArgs e)
    {
      Cursor oldCursor;

      oldCursor = Cursor.Current;
      Cursor.Current = Cursors.WaitCursor;
      try
      {
        var signHelper = new SignedRequestHelper(Amazon.AccessKey, Amazon.SecretKey, "ecs.amazonaws.com");
        var parameters = new Dictionary<String, String>() {
          { "Service", Amazon.ServiceName},
          { "Version", Amazon.ServiceVersion },
          { "Operation", "ItemSearch" },
          { "SearchIndex", "Books" },
          { "ResponseGroup", "Large" },
          { "Keywords", keywords.Text }
        };

        var url = signHelper.Sign(parameters);
        amazonXml = XElement.Load(url);

        var books = from result in amazonXml.Descendants(ns + "Item")
                    let attributes = result.Element(ns + "ItemAttributes")
                    select new Book
                           {
                             Isbn = (string)attributes.Element(ns + "ISBN"),
                             Title = (string)attributes.Element(ns + "Title"),
                           };

        bookBindingSource.DataSource = books;
      }
      finally
      {
        Cursor.Current = oldCursor;
      }
    }

    void importButton_Click(object sender, EventArgs e)
    {
      var selectedBooks =
        from row in bookDataGridView.Rows.OfType<DataGridViewRow>()
        where (bool)row.Cells[0].EditedFormattedValue
        select (Book)row.DataBoundItem;

      var booksToImport =
        from amazonItem in amazonXml.Descendants(ns + "Item")
        join selectedBook in selectedBooks
          on (string)amazonItem.Element(ns + "ItemAttributes").Element(ns + "ISBN")
          equals selectedBook.Isbn
        join p in ctx.Publishers
          on (string)amazonItem.Element(ns + "ItemAttributes").Element(ns + "Publisher")
          equals p.Name
          into publishers
        from existingPublisher in publishers.DefaultIfEmpty()
        let attributes = amazonItem.Element(ns + "ItemAttributes")
        select new Book
               {
                 ID = Guid.NewGuid(),
                 Isbn = (string)attributes.Element(ns + "ISBN"),
                 Title = (string)attributes.Element(ns + "Title"),
                 PublisherObject = (existingPublisher ??
                            new Publisher
                            {
                              ID = Guid.NewGuid(),
                              Name = (string)attributes.Element(ns + "Publisher")
                            }
                 ),
                 SubjectObject = (Subject)categoryComboBox.SelectedItem,
                 PubDate = (DateTime)attributes.Element(ns + "PublicationDate"),
                 Price = ParsePrice(attributes.Element(ns + "ListPrice")),
                 BookAuthors = GetAuthors(attributes.Elements(ns + "Author"))
               };

      foreach (Book book in booksToImport)
      {
        ctx.Books.InsertOnSubmit(book);
      }

      try
      {
        ctx.SubmitChanges();
        MessageBox.Show(booksToImport.Count() + " books imported.");
      }
      catch (Exception ex)
      {
        MessageBox.Show("An error occured while attempting to import the selected " +
          "books. " + Environment.NewLine + Environment.NewLine + ex.Message);
      }
    }

    private EntitySet<BookAuthor> GetAuthors(IEnumerable<XElement> authorElements)
    {
      var bookAuthors =
        from authorElement in authorElements
        join a in ctx.Authors on (string)authorElement equals a.FirstName + " " +
            a.LastName into authors
        from existingAuthor in authors.DefaultIfEmpty()
        let nameParts = ((string)authorElement).Split(' ')
        select new BookAuthor
               {
                 AuthorObject = existingAuthor ??
                 new Author
                 {
                   ID = Guid.NewGuid(),
                   FirstName = nameParts[0],
                   LastName = nameParts[1]
                 }
               };

      EntitySet<BookAuthor> set = new EntitySet<BookAuthor>();
      set.AddRange(bookAuthors);
      return set;
    }

    private decimal ParsePrice(XElement priceElement)
    {
      return Convert.ToDecimal(((string)priceElement.Element(ns +
               "FormattedPrice")).Replace("$", String.Empty));
    }
  }
}

File: Language/C#/Book/LINQ实战/第13章 应用程序各层中的LINQ/Chapter13.VB - LinqBooks/LinqBooks.ImportFromAmazon/ImportForm.vb
Imports System.Data.Linq
Imports LinqBooks.Entities
Imports AmazonCommon

Public Class ImportForm

  Dim ctx As LinqBooksDataContext = New LinqBooksDataContext(My.MySettings.Default.liaConnectionString)
  Dim ns As XNamespace = Amazon.AmazonNS
  Dim amazonXml As XElement

  Private Sub ImportForm_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
    categoryComboBox.DataSource = ctx.Subjects.OrderBy(Function(subject) subject.Name).ToList()
  End Sub

  Private Sub searchButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles searchButton.Click
    Dim oldCursor As Cursor

    oldCursor = Cursor.Current
    Cursor.Current = Cursors.WaitCursor
    Try
      Dim signHelper = New SignedRequestHelper(Amazon.AccessKey, Amazon.SecretKey, "ecs.amazonaws.com")
      Dim parameters = New Dictionary(Of String, String)
      parameters("Service") = Amazon.ServiceName
      parameters("Version") = Amazon.ServiceVersion
      parameters("Operation") = "ItemSearch"
      parameters("SearchIndex") = "Books"
      parameters("ResponseGroup") = "Large"
      parameters("Keywords") = keywords.Text

      Dim url = signHelper.Sign(parameters)
      amazonXml = XElement.Load(url)

      Dim books = From result In amazonXml.Descendants(ns + "Item") _
                  Let attributes = result.Element(ns + "ItemAttributes") _
                  Select New Book With _
                         { _
                           .Isbn = CType(attributes.Element(ns + "ISBN"), String), _
                           .Title = CType(attributes.Element(ns + "Title"), String) _
                         }

      bookBindingSource.DataSource = books
    Finally
      Cursor.Current = oldCursor
    End Try

  End Sub

  Private Sub importButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles importButton.Click
    Dim selectedBooks = _
      From row In bookDataGridView.Rows.OfType(Of DataGridViewRow)() _
      Where CType(row.Cells(0).EditedFormattedValue, Boolean) _
      Select CType(row.DataBoundItem, Book)

    Dim booksToImport = _
      From amazonItem In amazonXml.Descendants(ns + "Item") _
        Join selectedBook In selectedBooks _
          On CType(amazonItem.Element(ns + "ItemAttributes").Element(ns + "ISBN"), String) Equals selectedBook.Isbn _
        Group Join p In ctx.Publishers _
          On CType(amazonItem.Element(ns + "ItemAttributes").Element(ns + "Publisher"), String) Equals p.Name Into publishers = Group _
        From existingPublisher In publishers.DefaultIfEmpty() _
        Let attributes = amazonItem.Element(ns + "ItemAttributes") _
        Select New Book With _
               { _
                 .ID = Guid.NewGuid(), _
                 .Isbn = CType(attributes.Element(ns + "ISBN"), String), _
                 .Title = CType(attributes.Element(ns + "Title"), String), _
                 .PublisherObject = If(Not existingPublisher Is Nothing, existingPublisher, _
                            New Publisher With _
                            { _
                              .ID = Guid.NewGuid(), _
                              .Name = CType(attributes.Element(ns + "Publisher"), String) _
                            } _
                 ), _
                 .SubjectObject = CType(categoryComboBox.SelectedItem, Subject), _
                 .PubDate = CType(attributes.Element(ns + "PublicationDate"), DateTime), _
                 .Price = ParsePrice(attributes.Element(ns + "ListPrice")), _
                 .BookAuthors = GetAuthors(attributes.Elements(ns + "Author")) _
               }

    For Each book In booksToImport
      ctx.Books.InsertOnSubmit(book)
    Next

    Try
      ctx.SubmitChanges()
      MsgBox(booksToImport.Count() & " books imported.")
    Catch ex As Exception
      MsgBox("An error occured while attempting to import the selected " & _
        "books. " & Environment.NewLine & Environment.NewLine & ex.Message)
    End Try
  End Sub

  Private Function GetAuthors(ByVal authorElements As IEnumerable(Of XElement)) As EntitySet(Of BookAuthor)
    Dim bookAuthors = _
      From authorElement In authorElements _
      Group Join a In ctx.Authors On CType(authorElement, String) Equals a.FirstName & " " & a.LastName Into authors = Group _
      From existingAuthor In authors.DefaultIfEmpty() _
      Let nameParts = CType(authorElement, String).Split(" ") _
      Select New BookAuthor With _
             { _
               .AuthorObject = If(Not existingAuthor Is Nothing, existingAuthor, _
                   New Author With _
                   { _
                     .ID = Guid.NewGuid(), _
                     .FirstName = nameParts(0), _
                     .LastName = nameParts(1) _
                   }) _
             }

    Dim result = New EntitySet(Of BookAuthor)()
    result.AddRange(bookAuthors)
    Return result
  End Function

  Private Function ParsePrice(ByVal priceElement As XElement) As Decimal
    Dim formattedPrice = CType(priceElement.Element(ns + "FormattedPrice"), String)
    Return Convert.ToDecimal(formattedPrice.Replace("$", String.Empty))
  End Function

End Class